### DISCUSSION TITLE
IDL stubs generated with C++11 support for Java bindings

### DISCUSSION THREAD

**Discussion by valerijbielskij** [OpenDDS/OpenDDS#3876] [Q&A] [answered]
Hello DDS enthusiasts, I have a question related to Java bindings. Is it possible to generate IDL based stubs with C++11 support? Following Developers Guide 10.3 I've tried `dcps_ts_flags += -Wb,export_macro=Foo_Export -Lc++11` which yields lots of compilation errors during various stages of build process.
Our main code base is already relying on C++11 support enabled, now we are introducing a separate client (Java app) which will use JNI to pass objects in both directions, it would be great if we could support same C++ types both in the native C++ apps and the Java app. Please note that we are interested only in generated Java classes for the stubs, using Java based pub/sub implementations (or rather whole DCPS setup) is not an option as it would require reimplementing huge amount of communication logic back in Java.

**Comment by mitza-oci** [comment, marked as answer]
It isn't currently supported to have Java bindings layered on top of the C++11 language mapping.  The code generation for JNI assumes that the definitions from the "C++ classic" mapping are available.

Could you describe some more about how the Java application is set up?  OpenDDS is designed for the Java application to invoke the native code that it generates.  It's not designed for the application to provide its own native code.

It's not a problem to use these tools in different ways than they're designed (after all, this is open source!) but there may be some additional changes needed to support this use case.

**Reply by valerijbielskij** [reply]
> It's not designed for the application to provide its own native code.

That's exactly our case. Imagine that you have C++ clients which pass around objects generated from IDL, these clients are built using a common library which uses OpenDDS as a transport but has a thick layer of rules on how/when/from whom to receive these objects and perform more additional operations. Now there is a need for the Java service which operates on these objects using exactly same rules as C++ clients. 
We have implemented JNI bridge which loads aforementioned C++ library and uses `copyToJava` and `copyToCxx` functions for seamless conversions between Java and C++. For now we've introduced manual conversions between classic and C++11 mappings in the JNI bridge but we would like to get rid of it.
If someone could draw the guidelines on how to extend the Java bindings generator we could consider implementing it.

**Reply by mitza-oci** [reply]
I think it's feasible to implement a command-line option in `idlj2ni` which would cause it generate code for the C++11 binding.  That option needs to be enabled only when `idl2jni` processes the user-written IDL and not when it processes the generated `*TypeSupport.idl`.

To see how this would work, start with an example like `java/tests/messenger/messenger_idl/MessengerJC.cpp`.  The generated code currently accesses the data members of class `Messenger::Message` directly.  For example, the object `target` in:
```
    copyToCxx (jni, target.from, obj);
```

In the C++11 mapping this would change to either
```
    copyToCxx (jni, target.from(), obj);
```
or
```
    copyToCxx (jni, target._from, obj);
```
(the latter is non-standard so it may need to change if opendds_idl changes how it implements the IDL-to-C++11 Language Mapping spec)

Some of the supporting functions need to be extended as well, like `std::string` versions of `copyToCxx` and `copyToJava` in `idl2jni_runtime.h`.

**Reply by valerijbielskij** [reply]
Thanks a lot for the suggestions, indeed type support classes can be skipped making it an affordable effort, we will see how it goes.
